File: packages/sdk-typescript/tests/unit/team.test.ts
// V-309e — TeamResource SDK tests.

import { describe, expect, it, vi } from 'vitest';
import { TeamResource } from '../../src/resources/team.js';
import type { HttpClient } from '../../src/http.js';

interface RequestOpts {
  method: string;
  path: string;
  body?: unknown;
}

describe('TeamResource', () => {
  it('invite POSTs /v1/team/invites with email + role', async () => {
    const calls: RequestOpts[] = [];
    const request = vi.fn((opts: RequestOpts) => {
      calls.push(opts);
      return Promise.resolve({ message: 'Invite sent.' });
    });
    const team = new TeamResource({ request } as unknown as HttpClient);
    await team.invite('user@example.test', { role: 'admin' });
    expect(calls).toHaveLength(1);
    expect(calls[0]!.method).toBe('POST');
    expect(calls[0]!.path).toBe('/v1/team/invites');
    expect(calls[0]!.body).toEqual({ email: 'user@example.test', role: 'admin' });
  });

  it('invite without role omits the field', async () => {
    const calls: RequestOpts[] = [];
    const request = vi.fn((opts: RequestOpts) => {
      calls.push(opts);
      return Promise.resolve({ message: 'Invite sent.' });
    });
    const team = new TeamResource({ request } as unknown as HttpClient);
    await team.invite('user@example.test');
    expect(calls[0]!.body).toEqual({ email: 'user@example.test' });
  });

  it('listMembers GETs /v1/team/members', async () => {
    const request = vi.fn(() => Promise.resolve({ data: [] }));
    const team = new TeamResource({ request } as unknown as HttpClient);
    const result = await team.listMembers();
    expect(result.data).toEqual([]);
    expect(request).toHaveBeenCalledWith({ method: 'GET', path: '/v1/team/members' });
  });

  it('listInvites GETs /v1/team/invites', async () => {
    const request = vi.fn(() => Promise.resolve({ data: [] }));
    const team = new TeamResource({ request } as unknown as HttpClient);
    await team.listInvites();
    expect(request).toHaveBeenCalledWith({ method: 'GET', path: '/v1/team/invites' });
  });

  it('listOwners GETs /v1/team/owners with a typed workspace envelope', async () => {
    const request = vi.fn(() =>
      Promise.resolve({
        data: [
          {
            owner_account_id: 'acc_owner',
            owner_email: 'owner@example.test',
            owner_name: 'Owner',
            role: 'admin',
            membership_id: 'mem_test',
          },
        ],
      }),
    );
    const team = new TeamResource({ request } as unknown as HttpClient);
    const result = await team.listOwners();
    expect(result.data[0]?.owner_email).toBe('owner@example.test');
    expect(request).toHaveBeenCalledWith({ method: 'GET', path: '/v1/team/owners' });
  });

  it('acceptInvite POSTs token in body', async () => {
    const calls: RequestOpts[] = [];
    const request = vi.fn((opts: RequestOpts) => {
      calls.push(opts);
      return Promise.resolve({
        membership: {
          id: 'mem_test',
          owner_account_id: 'acc_owner',
          member_account_id: 'acc_member',
          member_email: 'm@example.test',
          role: 'member',
          invited_at: '2026-05-08T00:00:00Z',
          accepted_at: '2026-05-08T00:00:00Z',
          invited_by_account_id: 'acc_owner',
        },
      });
    });
    const team = new TeamResource({ request } as unknown as HttpClient);
    const result = await team.acceptInvite('plaintexttoken');
    expect(calls[0]!.path).toBe('/v1/team/invites/accept');
    expect(calls[0]!.body).toEqual({ token: 'plaintexttoken' });
    expect(result.membership.member_email).toBe('m@example.test');
  });

  it('removeMember DELETEs the membership URL', async () => {
    const calls: RequestOpts[] = [];
    const request = vi.fn((opts: RequestOpts) => {
      calls.push(opts);
      return Promise.resolve(undefined);
    });
    const team = new TeamResource({ request } as unknown as HttpClient);
    await team.removeMember('mem_test');
    expect(calls[0]!.method).toBe('DELETE');
    expect(calls[0]!.path).toBe('/v1/team/members/mem_test');
  });

  // ⛔ listTeams and renameTeam shipped with NO test in ANY of the three SDKs —
  // not TypeScript, not Python, not Go (V-1978). Both target real routes with
  // matching verbs, so nothing was broken; nothing was checking either.

  it('listTeams GETs /v1/teams and returns the envelope verbatim', async () => {
    const team1 = {
      id: 'team_11111111-2222-3333-4444-555555555555',
      name: 'Acme',
      slug: null,
      owner_account_id: 'acc_1',
      created_at: '2026-05-01T00:00:00Z',
      updated_at: '2026-05-01T00:00:00Z',
    };
    const calls: RequestOpts[] = [];
    const request = vi.fn((opts: RequestOpts) => {
      calls.push(opts);
      return Promise.resolve({ data: [team1] });
    });
    const team = new TeamResource({ request } as unknown as HttpClient);
    const out = await team.listTeams();
    expect(calls).toHaveLength(1);
    expect(calls[0]!.method).toBe('GET');
    expect(calls[0]!.path).toBe('/v1/teams');
    // `slug` is published but always null today; a client that dropped it would
    // change the shape customers destructure.
    expect(out).toEqual({ data: [team1] });
  });

  it('listTeams sends NO body — it is a read, and a body would change the request', async () => {
    const calls: RequestOpts[] = [];
    const request = vi.fn((opts: RequestOpts) => {
      calls.push(opts);
      return Promise.resolve({ data: [] });
    });
    const team = new TeamResource({ request } as unknown as HttpClient);
    await team.listTeams();
    expect(calls[0]!.body).toBeUndefined();
  });

  it('CRITICAL renameTeam PATCHes /v1/teams/:id with exactly { name }', async () => {
    const calls: RequestOpts[] = [];
    const request = vi.fn((opts: RequestOpts) => {
      calls.push(opts);
      return Promise.resolve({ team: { id: 'team_1', name: 'New' } });
    });
    const team = new TeamResource({ request } as unknown as HttpClient);
    await team.renameTeam('team_1', 'New');
    expect(calls[0]!.method).toBe('PATCH');
    expect(calls[0]!.path).toBe('/v1/teams/team_1');
    // Exactly { name }: the route runs reportUnknownRequestFields against
    // RenameTeamBodySchema, so an extra key is a warning the caller never sees.
    expect(calls[0]!.body).toEqual({ name: 'New' });
  });

  it('CRITICAL renameTeam URL-encodes the team id, so a caller-supplied id cannot alter the path', async () => {
    const calls: RequestOpts[] = [];
    const request = vi.fn((opts: RequestOpts) => {
      calls.push(opts);
      return Promise.resolve({ team: { id: 'x', name: 'n' } });
    });
    const team = new TeamResource({ request } as unknown as HttpClient);
    await team.renameTeam('team/with space', 'n');
    expect(calls[0]!.path).toBe('/v1/teams/team%2Fwith%20space');
  });
});

File: packages/sdk-go/team_test.go
package driftstack

import (
	"context"
	"encoding/json"
	"net/http"
	"testing"
	"time"
)

// V-309g — TeamResource tests.

func TestTeam_Invite(t *testing.T) {
	t.Parallel()
	_, client := newServer(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/team/invites" || r.Method != "POST" {
			t.Errorf("unexpected request: %s %s", r.Method, r.URL.Path)
		}
		var body TeamInviteRequest
		if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
			t.Fatal(err)
		}
		if body.Email != "user@example.test" || body.Role != TeamRoleAdmin {
			t.Errorf("unexpected body: %+v", body)
		}
		w.Header().Set("content-type", "application/json")
		_ = json.NewEncoder(w).Encode(TeamInviteResponse{Message: "Invite sent."})
	})

	got, err := client.Team.Invite(
		context.Background(),
		&TeamInviteRequest{Email: "user@example.test", Role: TeamRoleAdmin},
	)
	if err != nil {
		t.Fatal(err)
	}
	if got.Message != "Invite sent." {
		t.Errorf("message=%q", got.Message)
	}
}

func TestTeam_AcceptInvite(t *testing.T) {
	t.Parallel()
	now := time.Now().UTC().Truncate(time.Second)
	_, client := newServer(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/team/invites/accept" || r.Method != "POST" {
			t.Errorf("unexpected request: %s %s", r.Method, r.URL.Path)
		}
		var body TeamAcceptRequest
		if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
			t.Fatal(err)
		}
		if body.Token != "plaintext" {
			t.Errorf("token=%q", body.Token)
		}
		w.Header().Set("content-type", "application/json")
		_ = json.NewEncoder(w).Encode(TeamAcceptResponse{
			Membership: TeamMember{
				ID:              "mem_test",
				OwnerAccountID:  "acc_owner",
				MemberAccountID: "acc_member",
				MemberEmail:     "m@example.test",
				Role:            TeamRoleMember,
				InvitedAt:       now,
				AcceptedAt:      now,
			},
		})
	})

	got, err := client.Team.AcceptInvite(context.Background(), "plaintext")
	if err != nil {
		t.Fatal(err)
	}
	if got.Membership.MemberEmail != "m@example.test" {
		t.Errorf("member_email=%q", got.Membership.MemberEmail)
	}
}

func TestTeam_ListOwners(t *testing.T) {
	t.Parallel()
	ownerName := "Workspace owner"
	_, client := newServer(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/team/owners" || r.Method != "GET" {
			t.Errorf("unexpected request: %s %s", r.Method, r.URL.Path)
		}
		w.Header().Set("content-type", "application/json")
		_ = json.NewEncoder(w).Encode(TeamOwnersList{Data: []TeamOwner{{
			OwnerAccountID: "acc_owner",
			OwnerEmail:     "owner@example.test",
			OwnerName:      &ownerName,
			Role:           TeamRoleAdmin,
			MembershipID:   "mem_test",
		}}})
	})

	got, err := client.Team.ListOwners(context.Background())
	if err != nil {
		t.Fatal(err)
	}
	if len(got.Data) != 1 || got.Data[0].OwnerName == nil || *got.Data[0].OwnerName != ownerName {
		t.Fatalf("owners=%+v", got.Data)
	}
}

func TestTeam_RemoveMember(t *testing.T) {
	t.Parallel()
	_, client := newServer(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/team/members/mem_test" || r.Method != "DELETE" {
			t.Errorf("unexpected request: %s %s", r.Method, r.URL.Path)
		}
		w.WriteHeader(204)
	})

	if err := client.Team.RemoveMember(context.Background(), "mem_test"); err != nil {
		t.Fatal(err)
	}
}

// ListTeams and RenameTeam shipped with NO test in ANY of the three SDKs
// (V-1978). Both target real routes with matching verbs, so nothing was broken;
// nothing was checking either.
func TestTeam_ListTeams(t *testing.T) {
	t.Parallel()
	var method, path string
	_, client := newServer(t, func(w http.ResponseWriter, r *http.Request) {
		method, path = r.Method, r.URL.Path
		w.Header().Set("content-type", "application/json")
		_ = json.NewEncoder(w).Encode(map[string]any{
			"data": []map[string]any{{
				"id":               "team_1",
				"name":             "Acme",
				"slug":             nil,
				"owner_account_id": "acc_1",
				"created_at":       "2026-05-08T10:00:00Z",
				"updated_at":       "2026-05-08T10:00:00Z",
			}},
		})
	})
	out, err := client.Team.ListTeams(context.Background())
	if err != nil {
		t.Fatal(err)
	}
	if method != "GET" || path != "/v1/teams" {
		t.Errorf("got %s %s, want GET /v1/teams", method, path)
	}
	if len(out.Data) != 1 || out.Data[0].Name != "Acme" {
		t.Fatalf("data=%+v", out.Data)
	}
	// Slug is published but always null today; decoding it as a nil *string
	// rather than "" is what tells a caller it is unset.
	if out.Data[0].Slug != nil {
		t.Errorf("slug should decode to nil, got %v", *out.Data[0].Slug)
	}
}

func TestTeam_RenameTeam_SendsExactlyTheNameAndEncodesTheID(t *testing.T) {
	t.Parallel()
	var method, rawURI string
	var body map[string]any
	_, client := newServer(t, func(w http.ResponseWriter, r *http.Request) {
		// RequestURI, not URL.Path: URL.Path is already percent-DECODED, so
		// asserting on it cannot distinguish an encoded id from an unencoded one
		// and the arm would pass either way.
		method, rawURI = r.Method, r.RequestURI
		_ = json.NewDecoder(r.Body).Decode(&body)
		w.Header().Set("content-type", "application/json")
		_ = json.NewEncoder(w).Encode(map[string]any{
			"team": map[string]any{
				"id": "team_1", "name": "New", "slug": nil,
				"owner_account_id": "acc_1",
				"created_at":       "2026-05-08T10:00:00Z",
				"updated_at":       "2026-05-08T10:00:00Z",
			},
		})
	})

	// A caller-supplied id must not be able to alter the path.
	out, err := client.Team.RenameTeam(context.Background(), "team/with space", "New")
	if err != nil {
		t.Fatal(err)
	}
	if method != "PATCH" {
		t.Errorf("method=%s, want PATCH", method)
	}
	if rawURI != "/v1/teams/team%2Fwith%20space" {
		t.Errorf("raw request URI = %q, want the id percent-encoded", rawURI)
	}
	// Exactly {"name": ...}: the route reports unknown request fields against
	// RenameTeamBodySchema, so an extra key is a warning the caller never sees.
	if len(body) != 1 || body["name"] != "New" {
		t.Errorf("body=%v, want exactly {name:New}", body)
	}
	if out.Team.Name != "New" {
		t.Errorf("team.name=%q", out.Team.Name)
	}
}
